pkg/virt-controller/watch/vmi.go

```diff
@@ -307,26 +309,58 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 		}
 	}
 
+	hasWffcDataVolume := false
+	// there is no reason to check for waitForFirstConsumer is there are failed DV's
+	if !hasFailedDataVolume {
+		for _, dataVolume := range dataVolumes {
+			if dataVolume.Status.Phase == cdiv1.WaitForFirstConsumer {
+				hasWffcDataVolume = true
+			}
+		}
+	}
+
 	conditionManager := controller.NewVirtualMachineInstanceConditionManager()
 	vmiCopy := vmi.DeepCopy()
-	podExists := podExists(pod)
+	vmiPodExists := podExists(pod) && !isTempPod(pod)
+	tempPodExists := podExists(pod) && isTempPod(pod)
 
 	vmiCopy, err := c.setActivePods(vmiCopy)
 	if err != nil {
 		return fmt.Errorf("Error detecting vmi pods: %v", err)
 	}
 
 	switch {
-
 	case vmi.IsUnprocessed():
-		if podExists {
+		if vmiPodExists {
 			vmiCopy.Status.Phase = virtv1.Scheduling
 		} else if vmi.DeletionTimestamp != nil {
 			vmiCopy.Status.Phase = virtv1.Failed
 		} else if hasFailedDataVolume {
 			vmiCopy.Status.Phase = virtv1.Failed
 		} else {
 			vmiCopy.Status.Phase = virtv1.Pending
+			if hasWffcDataVolume {
+				if !tempPodExists {
+					condition := virtv1.VirtualMachineInstanceCondition{
+						Type:   virtv1.VirtualMachineInstanceProvisioning,
+						Status: k8sv1.ConditionTrue,
```

Comment:
wouldn't this `Provisioning` condition need to be set regardless if the temp pod exists or not? Otherwise there's no guarantee that the provisioning condition will get set. There's a race between setting this condition and creating the temp pod.

For instance, if update status fails to post the update to the VMI's status, then it's possible the next iteration that the temp pod exists and the provisioning condition is never set.

Reply:
you are right, it is simple, if WaitForFirstConsumer then controler needs to run special logic to force provisioning (provisioning condition set), then vmi transitions to Scheduling phase and controller can remove the condition!